### DISCUSSION TITLE
How to build small "native binaries without any dependencies"?

### DISCUSSION THREAD

**Discussion by parke** [vlang/v#19792] [Questions and Answers]
The [V homepage](https://vlang.io/) says:  "Performance - [V] Compiles to native binaries without any dependencies: a simple web server is only about 250 KB"

Are there instructions that explain how to build this 250 KB simple web server?

Or, more generally, are there instructions that explain how to build (hopefully small) native binaries without any dependencies?

**Comment by squidink7** [comment]
V compiles to small binaries effectively "without dependencies" by default, though that's not entirely accurate as they still depend on libc (which is already installed on 99.9% of operating systems). If you want an entirely statically linked binary that can run without even libc, you can pass `-cflags '-static'` to the v compiler. This will increase the size a bit as you're including a whole lot of functions that would normally be offered by the OS and its libraries, so you can shrink the size by using `-cflags '-Os -flto -static'` to perform extra optimizations.
For the website, I just tried building a simple hello world server with vweb and it came to about 840 KB, which while quite far from the 250 mentioned, is still not bad.

**Comment by medvednikov** [comment]
What size are you getting? Did you compile with -prod and -d use_openssl?

**Comment by medvednikov** [comment]
libc is not really a dependency, it's a system API

only linux supports binaries that are not linked to libc, you can use `v -cc musl-gcc foo.v` to build such binaries on linux

**Reply by parke** [reply]
> What size are you getting? Did you compile with -prod and -d use_openssl?

Are you asking me or @squidink7?

My hello world program (which was using the C `write()` function from `unistd.h`) was 764K.  I neither knew about nor used `-prod`, which is why I asked what the instructions were for building small binaries.

> libc is not really a dependency, it's a system API

I consider a dynamically loaded libc to be a dependency.  Some Linux distros use musl instead of glibc.  Binaries that expect a dynamically loaded glibc won't run if only musl is available on the host.  Also, I believe that sometimes binaries compiled against a newer glibc will not run with an older glibc (but I could be wrong about this).

Thanks for the tip about `v -cc musl-gcc foo.v`.

**Reply by JalonSolov** [reply]
You are correct that binaries compiled against newer glibc won't (always) work with older glibc.

`-prod` option basically forces V to use whichever C compiler you have installed _other than_ tcc, and turns on optimization.

Try `v help build` and `v help build-c` for other options.

**Reply by medvednikov** [reply]
Libc is not a dependency on anything but Linux, like I said. Arguably maybe even on musl flavors only.
On BSDs, macOS it's the only official API, on Windows it's also kind of an official API.
You can't avoid libc on these systems, the syscalls are not stable across different OS versions.

**Comment by justinas** [comment]
I don't think the homepage should claim that by default (e.g. just compiling with `v .`, without any additional flags) one can:

> <...> get a single statically linked binary <...>

Because by default, binaries are dynamically linked.

It is a fair point that `libc` is the only stable system API on many OSes, and I agree that is it a practical choice to use libc. Yet, it does not change the fact that the default output of the V compiler (with both the C backend, and the native codegen backend) is a dynamically-linked executable.

**Comment by abcxcba** [comment]
I took "**Compiles to native binaries without any dependencies**" literally, until I read a good answer: "**Libc is not a Libc is not a dependency on anything but Linux**", which showed me how badly I misunderstood "without any dependencies".

Yes, programs running on Windows do not depend on anything Linux, and programs running on Linux do not depend on anything plan9 ... ... 

It's great that there are no dependencies in the world, let alone "hellish dependencies", and now I feel very relieved.

Thanks to the author of this reply.

BTW I always thought that the author of the odin language was too far from the truth about v. Now I agree with him, but not with his point of view.

**Reply by JalonSolov** [reply]
Programs on Windows don't depend on anything from Linux, but they _do_ depend on the Windows system DLLs.  Yet nobody claims those as dependencies, either.

**Comment by badboycxcc** [comment]
Oh, why is an http.get request over 4MB?
```
PS D:\Code\V> v .\httpget.v -o 1.exe -os windows -pord -minimal
PS D:\Code\V> dir


    目录: D:\Code\V


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----        2024/10/21     14:05        4956672 1.exe
-a----        2024/10/21     14:04            237 httpget.v

```
```
PS D:\Code\V> v .\httpget.v -o 2.exe -os windows  -cflags '-Os -flto -static'
==================
C:/Users/Administrator/AppData/Local/Temp/v_0/2.exe.01JAPTVZS8BY0QCGY0NXFDSPNV.tmp.c:1225: warning: assignment from incompatible pointer type
tcc: error: library 'dbghelp' not found
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

**Comment by badboycxcc** [comment]
How to compile small, independent (no dependencies) native binaries？

**Comment by badboycxcc** [comment]
build 979KB 
```
D:\Code\V>v -prod -cflags -static httpget.v

D:\Code\V>
D:\Code\V>dir
2024/10/21  14:05         4,956,672 1.exe
2024/10/21  14:16           979,385 httpget.exe
2024/10/21  14:04               237 httpget.v
```

**Reply by Jengro777** [reply]
```
avey@avey vprod % v -prod -cflags -static .
================== C compilation error (from cc): ==============
cc: ld: library not found for -lcrt0.o
cc: clang: error: linker command failed with exit code 1 (use -v to see invocation)
... (the original output was 684 lines long, and was truncated to 2 lines)
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
avey@avey vprod %
```
macos

**Reply by jorgeluismireles** [reply]
This is a veb serving an building a small wasm (an index.html should be in the folder):
```v
import veb
import os

const http_port = 3001

struct Context {
	veb.Context
}

pub struct App {
	veb.StaticHandler
}

fn main() {
	os.chdir(os.dir(os.executable()))!
	mut app := &App{}
	veb.run_at[App, Context](mut app, port: http_port)!
}

pub fn (mut app App) before_accept_loop() {
	os.execute_or_panic('v -b js_browser draw.js.v ')
	app.handle_static('.', true) or { panic('${err}') }
}
```
I compile four times the code above:
```
v main.v -o main_default
v -cc gcc main.v -o main_gcc
v -cc gcc -prod main.v -o main_gcc_prod
v -cc gcc -prod -compress main.v -o main_gcc_prod_compress
```
These are the sizes: 4M, 2M, 1M and 360k
```
ls -la main_*
-rwxrwxr-x 1 jorge jorge 4517744 oct 21 09:23 main_default
-rwxrwxr-x 1 jorge jorge 2648368 oct 21 09:23 main_gcc
-rwxrwxr-x 1 jorge jorge 1034000 oct 21 09:24 main_gcc_prod
-rwxrwxr-x 1 jorge jorge  361496 oct 21 09:24 main_gcc_prod_compress
```
All they work correctly. For the later option `compress` v requested to have [upx](https://github.com/upx/upx) so I installed for ubuntu:
```
$ sudo apt install upx
```
Using `-skip-unused` crashed the compiler, which is another issue...
